src/ui/component/userSignIn/view.jsx

```diff
@@ -4,98 +4,132 @@ import { withRouter } from 'react-router';
+  const isCurrentlyFetchingSomething = fetchingChannels || claimingReward || syncingWallet;
+  const isWaitingForSomethingToFinish =
+    // If the user has claimed the email award, we need to wait until the balance updates sometime in the future
+    !hasFetchedReward || (hasFetchedReward && balance === 0) || (syncEnabled && !hasSynced);
+
+  // The possible screens for the sign in flow
+  const showEmail = !emailToVerify && !hasVerifiedEmail;
+  const showEmailVerification = emailToVerify && !hasVerifiedEmail;
+  const showUserVerification = hasVerifiedEmail && !rewardsApproved;
+  const showSyncPassword = syncEnabled && getSyncError && !hasSynced;
+  const showChannelCreation =
+    hasVerifiedEmail && balance && balance > DEFAULT_BID_FOR_FIRST_CHANNEL && channelCount === 0 && !hasYoutubeChannels;
+  const showYoutubeTransfer =
+    hasVerifiedEmail && hasYoutubeChannels && (hasTransferrableYoutubeChannels || hasPendingYoutubeTransfer);
+  const showLoadingSpinner =
+    canHijackSignInFlowWithSpinner && (isCurrentlyFetchingSomething || isWaitingForSomethingToFinish);
 
   React.useEffect(() => {
-    if (hasVerifiedEmail && balance !== undefined && !hasClaimedEmailAward && !hasFetchedReward) {
-      memoizedClaimReward();
+    fetchUser();
+  }, [fetchUser]);
+
+  React.useEffect(() => {
+    // Don't claim the reward if sync is enabled until after a sync has been completed successfully
+    // If we do it before, we could end up trying to sync a wallet with a non-zero balance which will fail to sync
+    const delayForSync = syncEnabled && !hasSynced;
+
+    if (hasVerifiedEmail && !hasClaimedEmailAward && !hasFetchedReward && !delayForSync) {
+      claimReward();
     }
-  }, [hasVerifiedEmail, memoizedClaimReward, balance, hasClaimedEmailAward, hasFetchedReward]);
+  }, [hasVerifiedEmail, claimReward, hasClaimedEmailAward, hasFetchedReward, syncEnabled, hasSynced, balance]);
 
-  if (
-    !user ||
-    (balance === 0 && !hasFetchedReward) ||
-    (hasVerifiedEmail && balance === undefined) ||
-    (hasVerifiedEmail && !hasFetchedChannels)
-  ) {
-    return (
+  // Loop through this list from the end, until it finds a matching component
+  // If it never finds one, assume the user has completed every step and redirect them
+  const SIGN_IN_FLOW = [
+    showEmail && <UserEmailNew />,
+    showEmailVerification && <UserEmailVerify />,
+    showUserVerification && <UserVerify />,
+    showSyncPassword && <SyncPassword />,
+    showChannelCreation && <UserFirstChannel />,
+    showYoutubeTransfer && <YoutubeTransferWelcome />,
+    showLoadingSpinner && (
       <div className="main--empty">
-        <Spinner delayed />
+        <Spinner />
       </div>
-    );
-  }
-
-  if (balance === 0 && hasClaimedEmailAward) {
-    history.replace(redirect || '/');
-  }
+    ),
+  ];
 
-  if (rewardsApproved && channelCount > 0) {
-    history.replace(redirect || '/');
+  let componentToRender;
+  for (var i = SIGN_IN_FLOW.length - 1; i > -1; i--) {
```

Comment:
Does this logic work if the state changes such that another step is added? I would assume it does since `SIGN_IN_FLOW` will refresh as well and you're always going to the first incomplete step...

Reply:
Yep